test/replication-luatest/gh_7590_replicaset_assert_test.lua

```diff
@@ -0,0 +1,93 @@
+        replication = {
+            server.build_listen_uri('r1'),
+            server.build_listen_uri('r3'),
+            server.build_listen_uri('r2'),
+        },
+    }
+
+    -- We need to specify uuid explicitly, so that rb tree traversal
+    -- wouldn't take errored applier (applier from r2 to r3) first.
+    g.cfg.instance_uuid = '00000000-0000-0000-0000-000000000001'
+    g.r1 = g.cluster:build_and_add_server({alias = 'r1', box_cfg = g.cfg})
+
+    g.cfg.instance_uuid = '00000000-0000-0000-0000-000000000002'
+    g.r2 = g.cluster:build_and_add_server({alias = 'r2', box_cfg = g.cfg})
+
+    g.cfg.instance_uuid = '00000000-0000-0000-0000-000000000003'
+    g.r3 = g.cluster:build_and_add_server({alias = 'r3', box_cfg = g.cfg})
+
+    g.cluster:start()
+    g.r2:exec(function()
+        box.schema.create_space('test'):create_index('pk')
+    end)
+
+    g.cluster:wait_for_fullmesh()
+    g.r3:wait_for_vclock_of(g.r2)
+    g.r1:wait_for_vclock_of(g.r2)
+end)
+
+g.after_all(function(g)
+    g.cluster:drop()
+end)
+
+g.test_replicaset_state_machine_assert_fail = function(g)
+    g.r1:exec(function(cfg)
+        -- Let's drop replication from r2 instance so that it cannot
+        -- trigger the destruction of applier from r1 to r3 at r3.
+        -- Heartbeats must be send, as we don't want the applier to
+        -- be destroyed on itself too.
+        cfg.replication[3] = nil
+        cfg.instance_uuid = nil
+        box.cfg(cfg)
+    end, {g.cfg})
+
+    g.r3:exec(function()
+        -- Throw ClientError as soon as write request from the r2 is received.
+        box.error.injection.set('ERRINJ_WAL_IO', true)
+        -- Don't allow applier_f fibers to be joined.
+        box.error.injection.set('ERRINJ_APPLIER_DESTROY_DELAY', true);
+        box.error.injection.set('ERRINJ_APPLIER_FAIL_DESTROY_DELAY', true);
+    end)
+
+    g.r2:exec(function()
+        -- Send some data via relay in order to trigger
+        -- destruction of applier from r2 to r3.
+        box.space.test:insert({1})
+    end)
+
+    t.helpers.retrying({timeout = 5}, function()
+        if not g.r3:grep_log('applier fail delay destroy') then
+            error("Applier destruction haven't started yet")
+        end
+    end)
+
+    g.r3:exec(function()
+        box.error.injection.set('ERRINJ_WAL_IO', false)
+        -- Errored applier can be waken up now. All others appliers will be
+        -- allowed to do the same as soon as errored one is unyielded
+        box.error.injection.set('ERRINJ_APPLIER_FAIL_DESTROY_DELAY', false);
+        require('fiber').create(function()
+            box.cfg({replication = {}})
+        end)
+    end)
+
+    t.helpers.retrying({timeout = 5}, function()
+        if not g.r3:grep_log('applier fail destroy continue') then
+            error("Applier destruction haven't continued yet")
+        end
+    end)
+
+    g.r3:exec(function()
+        box.error.injection.set('ERRINJ_APPLIER_DESTROY_DELAY', false);
```

Comment:
Let's unset "ERRINJ_WAL_IO" as well, just for consistency.

Reply:
It was done above, but all right, le's put it here, it doesn't affect anything